File: nim/src/gameboy.nim
import args
import cart
import ram
import cpu
import gpu
import apu
import buttons
import clock

type
    GameBoy* = object
        ram: ram.RAM
        cpu: cpu.CPU
        gpu: gpu.GPU
        apu: apu.APU
        buttons: buttons.Buttons
        clock: clock.Clock

proc create*(args: args.Args): GameBoy =
    let cart = cart.create(args.rom)
    let ram = ram.create(cart)
    let cpu = cpu.create(ram, args.debugCpu);
    let gpu = gpu.create(cpu, ram, cart.name, args.headless, args.debugGpu)
    let apu = apu.create(args.silent, args.debugApu)
    let buttons = buttons.create(cpu, ram, args.headless)
    let clock = clock.create(buttons, args.frames, args.profile, args.turbo)
    return GameBoy(
        ram: ram,
        cpu: cpu,
        gpu: gpu,
        apu: apu,
        buttons: buttons,
        clock: clock,
    )

proc tick*(gb: var GameBoy) =
    gb.cpu.tick()
    gb.gpu.tick()
    gb.buttons.tick()
    gb.clock.tick()
    gb.apu.tick()

proc run*(gb: var GameBoy) =
    while true:
        gb.tick()

File: go/src/gameboy.go
package main

type GameBoy struct {
	cart    *Cart
	ram     *RAM
	cpu     *CPU
	gpu     *GPU
	buttons *Buttons
	apu     *APU
	clock   *Clock
}

func NewGameBoy(args *Args) (*GameBoy, error) {
	cart, err := NewCart(args.Rom)
	if err != nil {
		return nil, err
	}
	ram := NewRAM(cart, args.DebugRam)
	cpu := NewCPU(&ram, args.DebugCpu)
	_, err = NewAPU(&cpu, args.DebugApu, args.Silent)
	if err != nil {
		return nil, err
	}
	gpu, err := NewGPU(&cpu, cart.name, args.DebugGpu, args.Headless)
	if err != nil {
		return nil, err
	}
	buttons, err := NewButtons(&cpu, args.Headless)
	if err != nil {
		return nil, err
	}
	clock, err := NewClock(buttons, args.Frames, args.Profile, args.Turbo)
	if err != nil {
		return nil, err
	}
	apu, err := NewAPU(&cpu, args.DebugApu, args.Silent)
	if err != nil {
		return nil, err
	}

	return &GameBoy{
		cart:    cart,
		ram:     &ram,
		cpu:     &cpu,
		gpu:     gpu,
		buttons: buttons,
		apu:     apu,
		clock:   clock,
	}, nil
}

func (gameboy *GameBoy) run() error {
	for {
		err := gameboy.tick()
		if err != nil {
			return err
		}
	}
}

func (gameboy *GameBoy) tick() error {
	err := gameboy.cpu.tick()
	if err != nil {
		return err
	}

	gameboy.gpu.tick()

	err = gameboy.buttons.tick()
	if err != nil {
		return err
	}

	err = gameboy.clock.tick()
	if err != nil {
		return err
	}

	gameboy.apu.tick()

	return nil
}
